README.md
# django-unslashed 

[![Version](https://pypip.in/version/django-unslashed/badge.png)](https://pypi.python.org/pypi/django-unslashed/) [![Build Status](https://travis-ci.org/dghubble/django-unslashed.png)](https://travis-ci.org/dghubble/django-unslashed) [![Downloads](https://pypip.in/download/django-unslashed/badge.png)](https://pypi.python.org/pypi/django-unslashed/) [![License](https://pypip.in/license/django-unslashed/badge.png)](https://pypi.python.org/pypi/django-unslashed/)

This middleware provides the inverse of the Django CommonMiddleware `APPEND_SLASH` feature. It can automatically remove trailing URL slashes and 301 redirect to the non-slash-terminated URL. This behavior is performed if the initial URL ends in a slash and is invalid, removing the trailing slash produces a valid URL, and `REMOVE_SLASH` is set to True. Otherwise there is no effect.

For example, foo.com/bar/ will be redirected to foo.com/bar if you don't have a valid URL pattern for foo.com/bar/ but do have a valid pattern for foo.com/bar and `REMOVE_SLASH=True`.


## Install

To install `django-unslashed`,

```
pip install django-unslashed
```

If you're using a `requirements.txt` file, add `django-unslashed>=0.3.0` to it.


## Usage

Modify your Django `settings.py` file to add `unslashed.middleware.RemoveSlashMiddleware`
to your `MIDDLEWARE_CLASSES` just before or after `django.middleware.common.CommonMiddleware`.

```python
MIDDLEWARE_CLASSES = (
    # ...
    'unslashed.middleware.RemoveSlashMiddleware',
    'django.middleware.common.CommonMiddleware',
    # ...
)
```

Set `REMOVE_SLASH` to True and `APPEND_SLASH` to False,

```python
APPEND_SLASH = False
REMOVE_SLASH = True
```

If `REMOVE_SLASH` is False or unset, the RemoveSlashMiddleware has no effect.


## Rationale

Web applications [should](http://googlewebmastercentral.blogspot.com/2010/04/to-slash-or-not-to-slash.html) have a URL structure which either:

1. Uses trailing slashes and redirects to append slashes if invalid non-slashed-terminated URLs are accessed.
2. Uses no trailing slash URLs and redirects to unslahed URLs if invalid slash terminated URLs are accessed. The prior is the Django default, while the later is possible by adding this middleware to your project.


## Notes

Based closely on Django's APPEND_SLASH CommonMiddleware [implementation](https://github.com/django/django/blob/master/django/middleware/common.py).


## Testing

```bash
$ git clone https://github.com/dghubble/django-unslashed.git
$ cd django-unslashed
$ python manage.py test unslashed
Creating test database for alias 'default'...
...
----------------------------------------------------------------------
Ran 6 tests in 0.071s

OK
Destroying test database for alias 'default'...
```


## License

[MIT License](LICENSE)
    


CHANGES.rst
Changelog
=========

Notable changes between releases.

0.3 (2014-12-18)
----------------

* Fix bug in which encoded characters were re-encoded upon redirection
* Add explicit test coverage for Django 1.6 and 1.7 and Python 2.6 to 3.4
* Require Django >= 1.5

0.2 (2013-12-18)
----------------

* Initial release

unslash/test.py
from django.http import HttpResponse, HttpResponsePermanentRedirect
from django.test import TestCase, Client

from django.middleware.common import CommonMiddleware
from unslash.middleware import UnslashMiddleware

class UnslashMiddlewareTest(TestCase):
  def setUp(self):
    self.client = Client()

  def test_permanent_redirect_to_non_slashed(self):
    response = self.client.get('/testapps/', follow=False)
    self.assertIsInstance(response, HttpResponsePermanentRedirect)
    self.assertFalse(response['Location'].endswith('/'))

    response = self.client.get('/testapps/1/', follow=False)
    self.assertIsInstance(response, HttpResponsePermanentRedirect)
    self.assertFalse(response['Location'].endswith('/'))

  def test_no_redirect_when_slash_url_is_valid(self):
    response = self.client.get('/testapps/1/urlendsinslash/', follow=False)
    self.assertIsInstance(response, HttpResponse)
    self.assertEqual(response.status_code, 200)
    
  def test_no_redirect_when_slashed_and_unslashed_invalid(self):
    response = self.client.get('/testapps/invalid/', follow=False)
    self.assertNotIsInstance(response, HttpResponsePermanentRedirect)

    response = self.client.get('/testapps/invalid', follow=False)
    self.assertNotIsInstance(response, HttpResponsePermanentRedirect)


  def tearDown(self):
    del self.client

unslashed/test.py
from django.http import HttpResponse, HttpResponsePermanentRedirect
from django.test import TestCase, Client
from django.utils.http import urlquote

from django.middleware.common import CommonMiddleware
from unslashed.middleware import RemoveSlashMiddleware

class RemoveSlashMiddlewareTest(TestCase):
    def setUp(self):
        self.client = Client()

    def test_permanent_redirect_to_non_slashed(self):
        response = self.client.get('/testapps/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))

        response = self.client.get('/testapps/1/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))

    def test_permanent_redirect_to_unslashed_when_url_has_plus_signs(self):
        response = self.client.get('/testapps/url+with+plus/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))
        self.assertRegexpMatches(response['Location'], r'/url\+with\+plus$')

        response = self.client.get('/testapps/url+with+plus/?param1=1&param2=%2Bfoo', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertRegexpMatches(response['Location'], r'/url\+with\+plus\?param1=1&param2=%2Bfoo$')

    def test_permanent_redirect_to_unslashed_when_url_has_urlencoded_chars(self):
        # urlquote is used because django.test.Client decodes the URL
        # which goes against the test
        slashed_url = urlquote('/testapps/quoted/foo%2Bbar%23baz%20/')
        response = self.client.get(slashed_url, follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))
        self.assertRegexpMatches(response['Location'], r'/quoted/foo%2Bbar%23baz%20$')

        response = self.client.get(slashed_url + '?param1=1&param2=%2Btest', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertRegexpMatches(response['Location'], r'/quoted/foo%2Bbar%23baz%20\?param1=1&param2=%2Btest$')

    def test_no_redirect_when_slash_url_is_valid(self):
        response = self.client.get('/testapps/1/urlendsinslash/', follow=False)
        self.assertIsInstance(response, HttpResponse)
        self.assertEqual(response.status_code, 200)

    def test_no_redirect_when_slashed_and_unslashed_invalid(self):
        response = self.client.get('/testapps/invalid/', follow=False)
        self.assertNotIsInstance(response, HttpResponsePermanentRedirect)
        self.assertEqual(response.status_code, 404)

        response = self.client.get('/testapps/invalid', follow=False)
        self.assertNotIsInstance(response, HttpResponsePermanentRedirect)
        self.assertEqual(response.status_code, 404)

    def test_warns_about_redirect_and_post(self):
        with self.settings(DEBUG=True):
            with self.assertRaises(RuntimeError):
                self.client.post('/testapps/', follow=False)

    def tearDown(self):
        del self.client


unslashed/tests.py
from django.http import HttpResponse, HttpResponsePermanentRedirect
from django.test import TestCase, Client
from django.utils.http import urlquote


class RemoveSlashMiddlewareTest(TestCase):
    def setUp(self):
        self.client = Client()

    def test_permanent_redirect_to_non_slashed(self):
        response = self.client.get('/testapps/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))

        response = self.client.get('/testapps/1/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))

    def test_permanent_redirect_to_unslashed_when_url_has_plus_signs(self):
        response = self.client.get('/testapps/url+with+plus/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))
        self.assertRegexpMatches(response['Location'], r'/url\+with\+plus$')

        response = self.client.get('/testapps/url+with+plus/?param1=1&param2=%2Bfoo', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertRegexpMatches(response['Location'], r'/url\+with\+plus\?param1=1&param2=%2Bfoo$')

    def test_permanent_redirect_to_unslashed_when_url_has_urlencoded_chars(self):
        # urlquote is used because django.test.Client decodes the URL
        # which goes against the test
        slashed_url = urlquote('/testapps/quoted/foo%2Bbar%23baz%20/')
        response = self.client.get(slashed_url, follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))
        self.assertRegexpMatches(response['Location'], r'/quoted/foo%2Bbar%23baz%20$')

        response = self.client.get(slashed_url + '?param1=1&param2=%2Btest', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertRegexpMatches(response['Location'], r'/quoted/foo%2Bbar%23baz%20\?param1=1&param2=%2Btest$')

    def test_no_redirect_when_slash_url_is_valid(self):
        response = self.client.get('/testapps/1/urlendsinslash/', follow=False)
        self.assertIsInstance(response, HttpResponse)
        self.assertEqual(response.status_code, 200)

    def test_no_redirect_when_slashed_and_unslashed_invalid(self):
        response = self.client.get('/testapps/invalid/', follow=False)
        self.assertNotIsInstance(response, HttpResponsePermanentRedirect)
        self.assertEqual(response.status_code, 404)

        response = self.client.get('/testapps/invalid', follow=False)
        self.assertNotIsInstance(response, HttpResponsePermanentRedirect)
        self.assertEqual(response.status_code, 404)

    def test_warns_about_redirect_and_post(self):
        with self.settings(DEBUG=True):
            with self.assertRaises(RuntimeError):
                self.client.post('/testapps/', follow=False)

    def tearDown(self):
        del self.client


unslashed/middleware.py
import re
from django import http
from django.conf import settings
from django.core import urlresolvers
from django.core.exceptions import ImproperlyConfigured


class RemoveSlashMiddleware(object):
    """
    This middleware provides the inverse of the APPEND_SLASH option built into
    django.middleware.common.CommonMiddleware. It should be placed just before
    or just after CommonMiddleware.

    If REMOVE_SLASH is True, the initial URL ends with a slash, and it is not
    found in the URLconf, then a new URL is formed by removing the slash at the
    end. If this new URL is found in the URLconf, then Django redirects the
    request to this new URL. Otherwise, the initial URL is processed as usual.

    For example, foo.com/bar/ will be redirected to foo.com/bar if you don't
    have a valid URL pattern for foo.com/bar/ but do have a valid pattern for
    foo.com/bar.

    Using this middlware with REMOVE_SLASH set to False or without REMOVE_SLASH
    set means it will do nothing.

    Orginally, based closely on Django's APPEND_SLASH CommonMiddleware
    implementation at
    https://github.com/django/django/blob/master/django/middleware/common.py.
    It has been reworked to use regular expressions instead of deconstructing/
    reconstructing the URL, which was problematically re-encoding some of the
    characters.
    """

    def process_request(self, request):

        old_url = request.build_absolute_uri()
        # match any / followed by ? (query string present)
        # OR match / at the end of the string (no query string)
        trailing_slash_regexp = r'(\/(?=\?))|(\/$)'
        new_url = old_url

        if getattr(settings, 'APPEND_SLASH') and getattr(settings, 'REMOVE_SLASH'):
            raise ImproperlyConfigured("APPEND_SLASH and REMOVE_SLASH may not both be True.")

        # Remove slash if REMOVE_SLASH is set and the URL has a trailing slash
        # and there is no pattern for the current path
        if getattr(settings, 'REMOVE_SLASH', False) and re.search(trailing_slash_regexp, old_url):
            urlconf = getattr(request, 'urlconf', None)
            if (not urlresolvers.is_valid_path(request.path_info, urlconf)) and urlresolvers.is_valid_path(request.path_info[:-1], urlconf):
                new_url = re.sub(trailing_slash_regexp, '', old_url)
                if settings.DEBUG and request.method == 'POST':
                    raise RuntimeError((""
                    "You called this URL via POST, but the URL ends in a "
                    "slash and you have REMOVE_SLASH set. Django can't "
                    "redirect to the non-slash URL while maintaining POST "
                    "data. Change your form to point to %s (without a "
                    "trailing slash), or set REMOVE_SLASH=False in your "
                    "Django settings.") % (new_url))

        if new_url == old_url:
            # No redirects required.
            return
        return http.HttpResponsePermanentRedirect(new_url)



testproject/settings.py
"""
Django settings for testproject project.

For more information on this file, see
https://docs.djangoproject.com/en/1.7/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.7/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
BASE_DIR = os.path.dirname(os.path.dirname(__file__))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.6/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '5foa=f)yc=^et#tboh(piigt(v^b+&8$pxvqcroid4le4_+1(%'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

TEMPLATE_DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'testapps',
)

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'unslashed.middleware.RemoveSlashMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'testproject.urls'

WSGI_APPLICATION = 'testproject.wsgi.application'

APPEND_SLASH = False

REMOVE_SLASH = True


# Database
# https://docs.djangoproject.com/en/1.6/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Internationalization
# https://docs.djangoproject.com/en/1.6/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.6/howto/static-files/

STATIC_URL = '/static/'
